### PR TITLE
Pin-aware SEE: correct pinners calculation and fix bug with pinned pieces giving check

### PR DESCRIPTION
**Pull request by pb00068**
This was a co-operation between Marco Costalba, Stéphane Nicolet and me.

Special thanks to Ronald de Man for reporting the bug with pinned pieces giving check,
discussed here: https://groups.google.com/forum/?fromgroups=#!topic/fishcooking/S_4E_Xs5HaE

STC:
LLR: 2.95 (-2.94,2.94) [-3.00,1.00]
Total: 132118 W: 23578 L: 23645 D: 84895

LTC:
LLR: 2.95 (-2.94,2.94) [-3.00,1.00]
Total: 36424 W: 4770 L: 4670 D: 26984

bench: 6272231

### CODE REVIEW

**Comment by pb00068** [comment]
P.S.: Along with renaming '**to**' to '**exchangeSquare**' I also propose to rename the variable '**captured**'.
While I had no problems to understand Ronalds example, in the begin I had big problems to understand the relative bug-fix-code, until recognizing that I wrongly interpreted '**captured**' in its passive form '**was captured**', while it's active form '**has captured**' must have been intended by the original author.
To avoid this ambiguity I suggest to either rename the variable into 'hasCaptured' or 'nextVictim'.

**Comment by snicolet** [comment]
*Snipers are sliders which point their rays towards the enemy king*

The square 's' is not always a king, it is a queen half of the time.

**Comment by pb00068** [comment]
@snicolet You are right! So maybe use Rocky's version: 
  _Snipers are 'sliders' that are aligned on target 's'_

**Comment by mcostalba** [comment]
Your interpretation is correct, it was intended passive form. The
misunderstanding arises because the variable is updated very early on, when
it is still the capturing one.  So perhaps next victim is better.

**Comment by pb00068** [comment]
'captured' is also a bit misleading in case of first move being a quiet move (nowadays we have a lot of such calls)

**Comment by Rocky640** [comment]
OK, let me do the devil's advocate. There are many ways to skin a cat, but this is new territory. 
I think we should proceed with caution in those cases.

Here are some numbers based on 2 instrumented bench run. This is not game playing conditions, but this gives some picture on what is going on.
 
One bench which was using the B see-value to continue the bench, and the other was using the C see-values.
Each time the SEE is called, I call also the other SEE versions, and count how many are different.

2016-09-19_15h12_36

Version B is more than 99.9 identical to version A, but was proven to be better using SPRT [0,5]
This was a great surprise, and quite a smart and efficient implementation.

Version C (this PR) does improve a few positions, but really, really not much as you can see.
I did not try to find if the few differences were due to Ronald's improvement or the changes in slider_blocker. If I have time I will analyse more and output the positions.

Version D refers to this ugly test of mine
http://tests.stockfishchess.org/tests/view/57df9a7b0ebc59763f358b1f
It detects exactly many more pinned situations, (I want manually into hundreds of them) but needs some optimisation, or there is possibly bug for a class of positions where it does not work. 
I will have to step back a few days before "see"ing clearly.

INVALID refers to early non-legal move detection that I added to my version D.
My thinking was that, if skipping (illegal) pinned moves in SEE is "better" why not go more generally and skip as many illegal moves as possible. But this is giving new responsibilities to SEE, and I don't have the general picture if on the long run this could be the right approach.

Hope this helps.

**Comment by pb00068** [comment]
@Rocky640 Thank you for your interesting work and analysis, appreciate it!
The goal of this PR wasn't to improve strength, it was to make the algorithm easier to understand (more self-explanatory) by getting rid of the vague term _potential pinners_

**Comment by snicolet** [comment]
Indeed the main benefit of the patch in this pull request is to clarify the semantics of the `pinners`parameter in slider_blockers()

• current master: `pinners` can return direct checkers or sliders with 1,2,3,... pieces (of either color) between the slider and the attacked square
• with this patch: `pinners` only returns sliders with exactly one defensive piece between the slider and the attacked square (in other words, `pinners` returns exact pinners).

The fact that the pull request patch does not lose Elo is a nice side effect.

In short, I think the best way forward is to apply the pull request, and try to build from it either by more Elo gaining patches or by simplifications, as usual.

**Comment by snicolet** [comment]
Having exact pinners in `pinners`also opens new opportunities in evaluation, btw.

**Comment by snicolet** [comment]
@mcostalba @pb00068

If **nextVictim** doesn't fit, then **moving** would perhaps be a good name to replace **captured** (but I like nextVictim too)

````
moving = type_of(piece_on(from));
....
moving = min_attacker<PAWN>(byTypeBB, to, stmAttackers, occupied, attackers)
....
if (   moving != KING
    && ..)
....
} while (stmAttackers && (moving != KING || (--slIndex, false)));
````

**Comment by mcostalba** [comment]
Merged with 943ae89be11929b

**Comment by pb00068** [comment]
@mcostalba  @snicolet 
The script which builds the abrok-development-builds page apparently does not like special characters: the credits to Stèphane Nicolet got simply truncated because of the 'è' !?

**Comment by snicolet** [comment]
é :-)

**Comment by pb00068** [comment]
Usually Marco's commits are reporting merely encoding: 'system' so it's not clear to me which encoding it is actually.

**Comment by mcostalba** [comment]
@pb00068 Could you please take a look at this?

https://groups.google.com/forum/?fromgroups=#!topic/fishcooking/S_4E_Xs5HaE

In particular to the last post of Bernhard Bauer.

**Comment by mcostalba** [comment]
@pb00068 sorry, I misread the post, indeed it is a bit odd but absolutely possible.

**Comment by pb00068** [comment]
@mcostalba Since I saw the result of this [sanity check](http://tests.stockfishchess.org/tests/view/57e2704b0ebc59763f358c32) by snicolet, I'm not so sure about this PR anymore.
Stephane suggested to test both fixes in one shot with the reasoning that it is even stricter (see his post in [forum-thread](https://groups.google.com/forum/?fromgroups=#!topic/fishcooking/S_4E_Xs5HaE)). Now if I interpret the result of his sanity check correctly, it means that with 95% certainty Ronald' fix wasn't just a cosmetic fix, it might be a true improvement in playing strength.
Under that light can we exclude the possibility that the correct-pinners-fix actually represents a regression but slipped in because of  Ronald' fix?

**Comment by pb00068** [comment]
@mcostalba To be on the safe side I made a [sanity check](http://tests.stockfishchess.org/tests/view/57e3e1100ebc59763f358cdb)
and everything seems just fine with this PR ☺
